src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,199 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
 
             var semanticVersion = await GetDocumentSemanticVersionAsync(documentSnapshot).ConfigureAwait(false);
 
-            // If we have a matching cached response, avoid computation and return early.
-            if (_cachedResponses.TryGetValue(textDocumentIdentifier.Uri, out var response) &&
-                response.IsCSharpFinalized &&
-                response.SemanticVersion == semanticVersion &&
-                response.Range == range)
+            // The LSP client shouldn't send us ranges that start or end midway through a line, but we'll be
+            // defensive here since our tokens cache is only meant to be used with complete lines.
+            if (range.Start.Character != 0 || range.End.Character != 0)
             {
-                return response.SemanticTokens;
+                _logger.LogWarning($"Requested range starts or ends midway through a line: {range}");
+                var (tokens, _) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, range, cancellationToken);
+
+                return tokens;
+            }
+
+            // See if we can use our cache to at least partially avoid recomputation.
+            if (!_tokensCache.TryGetCachedTokens(textDocumentIdentifier.Uri, semanticVersion, range, out var cachedResult))
+            {
+                // No cache results found, so we'll recompute tokens for the entire range and then hopefully cache the
+                // results to use next time around.
+                var (tokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, range, cancellationToken);
+                if (isCSharpFinalized && tokens is not null)
```

Comment:
what does isCSharpFinalized mean?

Reply:
`isCSharpFinalized` indicates to Razor whether Roslyn is returning tokens generated using an unfinalized semantic model (via frozen partial semantics). I'll make a note of it in the comments